File: multi/application/src/main/java/com/example/application/SeatReservation/redis/DistributedLockManager.java
package com.example.application.SeatReservation.redis;


import org.springframework.stereotype.Service;

@Service
public interface DistributedLockManager {
    <T> T executeWithLock(String key, long leaseTime, long waitTime, LockCallback<T> callback) throws Exception;

    @FunctionalInterface
    interface LockCallback<T> {
        T execute() throws Exception;
    }
}

File: module-common/src/main/kotlin/com/example/redis/managers/DistributedLockManager.kt
package com.example.redis.managers

import org.redisson.api.RLock
import org.redisson.api.RedissonClient
import org.springframework.stereotype.Component
import java.util.concurrent.TimeUnit

@Component
class DistributedLockManager(
    private val redissonClient: RedissonClient
) {
    fun <T> executeWithLock(
        lockKey: String,
        waitTime: Long = 0L,
        leaseTime: Long = 10L,
        timeUnit: TimeUnit = TimeUnit.SECONDS,
        task: () -> T
    ): T {
        val rLock: RLock = redissonClient.getLock(lockKey)
        return if(rLock.tryLock(waitTime, leaseTime, timeUnit)) {
            try {
                task()
            } finally {
                rLock.unlock()
            }
        } else {
            throw IllegalStateException()
        }
    }
}